### PR TITLE
[WIP] Query engine

### PR DESCRIPTION
**Pull request by luuse**
This implements fetching complete time series from Cassandra and an object structure for describing queries as java objects.

### CODE REVIEW

File: cassandra/src/main/java/se/tre/freki/storage/cassandra/query/DataPointIterator.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    @Override
+    public RowDataPoint dataPoint(final Row row) {
+      final RowDataPoint dataPoint = dataPointFor(row);
+      dataPoint.setRow(row);
+      dataPointIterator.typeStrategy = new StaticTypeStrategy(dataPoint);
+      return dataPoint;
+    }
+
+    private RowDataPoint dataPointFor(final Row row) {
```
Comment (Kasidro) [inline cassandra/src/main/java/se/tre/freki/storage/cassandra/query/DataPointIterator.java]: need to define types of datapoint (row) for float_value and double_value

File: cassandra/src/main/java/se/tre/freki/storage/cassandra/query/SpeculativePartitionIterator.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    return currentResultSet.one();
+  }
+
+  private void checkHasNext(final String message) {
+    if (!hasNext()) {
+      throw new NoSuchElementException(message);
+    }
+  }
+
+  private ResultSet nextPartitionResultSet() {
```
Comment (Kasidro) [inline cassandra/src/main/java/se/tre/freki/storage/cassandra/query/SpeculativePartitionIterator.java]: missing javadoc

File: cassandra/src/main/java/se/tre/freki/storage/cassandra/query/SpeculativePartitionIterator.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+      return fetchedResultSet.get();
+    } catch (ExecutionException e) {
+      throw new QueryException("Fetch of next partition threw an exception", e.getCause());
+    } catch (InterruptedException e) {
+      throw new QueryException("Interrupted while waiting for next partition");
+    }
+  }
+
+  private ListenableFuture<ResultSet> fetchNextPartition() {
```
Comment (Kasidro) [inline cassandra/src/main/java/se/tre/freki/storage/cassandra/query/SpeculativePartitionIterator.java]: missing javadoc

File: cassandra/src/main/java/se/tre/freki/storage/cassandra/query/DataPointIterator.java (hunk 1 above)
Comment (luuse) [inline cassandra/src/main/java/se/tre/freki/storage/cassandra/query/DataPointIterator.java]: This is fixed now.

File: core/src/main/java/se/tre/freki/query/predicate/AlternationTimeSeriesIdPredicate.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package se.tre.freki.query.predicate;
+
+import se.tre.freki.labels.LabelId;
+
+import com.google.common.collect.ImmutableSet;
+
+public class AlternationTimeSeriesIdPredicate extends TimeSeriesIdPredicate {
```
Comment (Kasidro) [inline core/src/main/java/se/tre/freki/query/predicate/AlternationTimeSeriesIdPredicate.java]: missing javadoc

File: core/src/main/java/se/tre/freki/query/predicate/SimpleTimeSeriesIdPredicate.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package se.tre.freki.query.predicate;
+
+import se.tre.freki.labels.LabelId;
+
+public class SimpleTimeSeriesIdPredicate extends TimeSeriesIdPredicate {
```
Comment (Kasidro) [inline core/src/main/java/se/tre/freki/query/predicate/SimpleTimeSeriesIdPredicate.java]: missing javadoc

File: core/src/main/java/se/tre/freki/query/predicate/TimeSeriesQueryPredicate.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import static com.google.common.base.Preconditions.checkState;
+
+import se.tre.freki.labels.LabelId;
+
+import com.google.common.collect.ImmutableSet;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class TimeSeriesQueryPredicate {
```
Comment (Kasidro) [inline core/src/main/java/se/tre/freki/query/predicate/TimeSeriesQueryPredicate.java]: missing javadoc

File: core/src/main/java/se/tre/freki/query/predicate/TimeSeriesQueryPredicate.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  private final LabelId metric;
+  private final ImmutableSet<TimeSeriesTagPredicate> tagPredicates;
+
+  protected TimeSeriesQueryPredicate(final LabelId metric,
```
Comment (Kasidro) [inline core/src/main/java/se/tre/freki/query/predicate/TimeSeriesQueryPredicate.java]: missing javadoc

File: core/src/main/java/se/tre/freki/query/predicate/TimeSeriesTagPredicate.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
+package se.tre.freki.query.predicate;
+
+public class TimeSeriesTagPredicate {
```
Comment (Kasidro) [inline core/src/main/java/se/tre/freki/query/predicate/TimeSeriesTagPredicate.java]: missing javadoc

File: core/src/main/java/se/tre/freki/query/predicate/WildcardTimeSeriesIdPredicate.java [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package se.tre.freki.query.predicate;
+
+public class WildcardTimeSeriesIdPredicate extends TimeSeriesIdPredicate {
```
Comment (Kasidro) [inline core/src/main/java/se/tre/freki/query/predicate/WildcardTimeSeriesIdPredicate.java]: mssing javadoc

File: core/src/main/java/se/tre/freki/query/predicate/TimeSeriesQueryPredicate.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    checkArgument(!tagPredicates.isEmpty());
+    this.tagPredicates = tagPredicates;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  private static class Builder {
```
Comment (Kasidro) [inline core/src/main/java/se/tre/freki/query/predicate/TimeSeriesQueryPredicate.java]: shokd this have javadoc?

File: core/src/main/java/se/tre/freki/query/TimeSeriesResultSet.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
+package se.tre.freki.query;
+
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Iterator;
+import javax.annotation.Nonnull;
+
+public interface TimeSeriesResultSet {
```
Comment (Kasidro) [inline core/src/main/java/se/tre/freki/query/TimeSeriesResultSet.java]: missing javadoc

File: core/src/main/java/se/tre/freki/query/DataPoint.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package se.tre.freki.query;
+
+public interface DataPoint {
```
Comment (Kasidro) [inline core/src/main/java/se/tre/freki/query/DataPoint.java]: missing javadoc

File: cassandra/src/main/java/se/tre/freki/storage/cassandra/query/DataPointIterator.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        return new RowDataPoint.RowFloatDataPoint();
+      } else if (columnDefinitions.contains("double_value")) {
+        return new RowDataPoint.RowDoubleDataPoint();
+      } else {
+        throw new AssertionError("row does not contain any known type");
+      }
+    }
+  }
+
+  private static class StaticTypeStrategy implements TypeStrategy {
```
Comment (Kasidro) [inline cassandra/src/main/java/se/tre/freki/storage/cassandra/query/DataPointIterator.java]: missing javadoc

**Pull request merged**